src/Magnum/SceneGraph/OctreeDrawableGroup.hpp

```diff
@@ -0,0 +1,113 @@
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included
+    in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+*/
+
+/** @file
+ * @brief @ref compilation-speedup-hpp "Template implementation" for @ref OctreeDrawableGroup.h
+ */
+
+#include <Magnum/Shapes/AbstractShape.h>
+#include <Magnum/Shapes/ShapeGroup.h>
+
+#include <Magnum/Octree/Octree.hpp>
+
+#include "Magnum/SceneGraph/OctreeDrawableGroup.h"
+
+namespace Magnum { namespace SceneGraph {
+
+using namespace Shapes;
+
+template<class T> OctreeDrawableGroup<T>& OctreeDrawableGroup<T>::add(Drawable<3, T>& drawable, const Range3D boundingBox) {
+    _drawables.push_back(&drawable);
+    _boundingBoxes.push_back(boundingBox);
+
+    if(_octree != nullptr) {
+        /* Destroy previous octree build */
+        delete _octree;
+    }
+
+    return *this;
+}
+
+template<class T> OctreeDrawableGroup<T>& OctreeDrawableGroup<T>::add(const DrawableGroup<3, T>& drawables, const std::vector<Range3D>& boundingBoxes) {
+    CORRADE_ASSERT(drawables.size() == boundingBoxes.size(), "Drawables and Shapes have to be equal size.", *this);
+    for(int i = 0; i < drawables.size(); ++i) {
+        _drawables.push_back(drawables[i]);
+        _boundingBoxes.push_back(boundingBoxes[i]);
+    }
+
+    if(_octree != nullptr) {
+        /* Destroy previous octree build */
+        delete _octree;
+    }
+
+    return *this;
+}
+
+template<class T> OctreeDrawableGroup<T>& OctreeDrawableGroup<T>::cull(Camera<3, T>& camera) {
+    CORRADE_ASSERT(_octree != nullptr, "Octree needs to be built before culling.", *this);
+
+    /* Clear before next cull */
+    for(UnsignedInt i = 0; i < this->size(); ++i) {
+        this->remove((*this)[i]);
+    }
+
+    /* Extract view frustum from view projection matrix */
+    const Matrix4 mvp = Matrix4(camera.projectionMatrix()*camera.cameraMatrix());
+    const Vector4 left{mvp.row(3) + mvp.row(0)};
+    const Vector4 right{mvp.row(3) - mvp.row(0)};
+    const Vector4 bottom{mvp.row(3) + mvp.row(1)};
+    const Vector4 top{mvp.row(3) - mvp.row(1)};
+    const Vector4 near{mvp.row(3) + mvp.row(2)};
+    const Vector4 far{mvp.row(3) - mvp.row(2)};
```

Comment:
Wasn't there a function for this somewhere? If not, can there be such a function? Or the `Frustum` class we were talking about?